### PR TITLE
[2.0] Rework Options

### PR DESCRIPTION
**Pull request by HazAT**
This brings all the options to the unified API.
The only option missing is `debug` but we do not have an internal debug logger yet, not really crucial for the first 2.0 release.

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -124,6 +124,10 @@ public function captureMessage(string $message, ?Severity $level = null, ?Scope
         $payload['message'] = $message;
         $payload['level'] = $level;
 
+        if ($this->getOptions()->getAttachStacktrace()) {
+            $payload['exception'] = new \Exception('Sentry Synthetic Exception');
```
Comment (Jean85) [inline src/Client.php]: What's this for?

File: src/Client.php [hunk 2]
```diff
@@ -264,16 +271,14 @@ private function addThrowableToEvent(Event $event, \Throwable $exception): void
...
-                $data['stacktrace'] = Stacktrace::createFromBacktrace(
-                    $this->getOptions(),
-                    $this->serializer,
-                    $this->representationSerializer,
-                    $currentException->getTrace(),
-                    $currentException->getFile(),
-                    $currentException->getLine()
-                );
-            }
+            $data['stacktrace'] = Stacktrace::createFromBacktrace(
```
Comment (Jean85) [inline src/Client.php]: Why did we loose the `if` check here?

File: tests/ClientBuilderTest.php [hunk 3]
```diff
@@ -188,13 +188,10 @@ public function optionsDataProvider(): array
...
             ['setSampleRate', 0.5],
             ['setMbDetectOrder', ['foo', 'bar']],
-            ['setAutoLogStacks', false],
+            ['setAttachStacktrace', true],
             ['setContextLines', 0],
             ['setEnableCompression', false],
-            ['setCurrentEnvironment', 'test'],
-            ['setEnvironments', ['default']],
-            ['setExcludedLoggers', ['foo', 'bar']],
-            ['setExcludedExceptions', ['foo', 'bar']],
```
Comment (Jean85) [inline tests/ClientBuilderTest.php]: You didn't remove `setExcludedExceptions`, that's an error?

File: src/ClientBuilder.php [hunk 4]
```diff
@@ -41,18 +41,12 @@
...
+ * @method setAttachStacktrace(bool $enable)
  * @method int getContextLines()
  * @method setContextLines(int $contextLines)
- * @method string getCurrentEnvironment()
- * @method setCurrentEnvironment(string $environment)
- * @method string[] getEnvironments()
- * @method setEnvironments(string[] $environments)
- * @method string[] getExcludedLoggers()
- * @method setExcludedLoggers(string[] $loggers)
- * @method string[] getExcludedExceptions()
```
Comment (Jean85) [inline src/ClientBuilder.php]: You didn't remove `get`/`setExcludedExceptions`

File: tests/ClientTest.php [hunk 5]
```diff
@@ -395,23 +389,6 @@ public function convertExceptionDataProvider()
                     ],
                 ],
             ],
-            [
-                new \RuntimeException('foo'),
-                [
-                    'auto_log_stacks' => false,
```
Comment (Jean85) [inline tests/ClientTest.php]: You're no longer testing for this case, why?

File: src/Options.php [hunk 6]
```diff
@@ -179,23 +179,23 @@ public function setMbDetectOrder($detectOrder): void
     }
 
     /**
-     * Gets whether the stacktrace must be auto-filled.
+     * Gets whether the stacktrace will be attached on captureMessage.
      *
      * @return bool
      */
-    public function getAutoLogStacks(): bool
+    public function getAttachStacktrace(): bool
```
Comment (ste93cry) [inline src/Options.php]: I would name this method `shouldAttachStacktrace` because it's not really a getter

File: src/Options.php [hunk 7]
```diff
@@ -179,23 +179,23 @@ public function setMbDetectOrder($detectOrder): void
     }
 
     /**
-     * Gets whether the stacktrace must be auto-filled.
+     * Gets whether the stacktrace will be attached on captureMessage.
```
Comment (ste93cry) [inline src/Options.php]: Is this option specifically designed to work only in the `captureMessage` method? It seems a bit strange...

File: src/ClientBuilder.php [hunk 8]
```diff
@@ -41,18 +41,12 @@
...
- * @method setExcludedExceptions(string[] $exceptions)
+ * @method null|string getEnvironment()
+ * @method setEnvironment(?string $environment)
```
Comment (ste93cry) [inline src/ClientBuilder.php]: The PHPDoc standard doesn't allow `nullable` values to be defined with the syntax PHP uses (the exclamation mark before the type). You always have to use the union of the type and the `null` value

File: src/Options.php [hunk 9]
```diff
@@ -179,23 +179,23 @@ public function setMbDetectOrder($detectOrder): void
...
-        return $this->options['auto_log_stacks'];
+        return $this->options['attach_stacktrace'];
     }
 
     /**
-     * Sets whether the stacktrace must be auto-filled.
+     * Sets whether the stacktrace will be attached on captureMessage.
      *
-     * @param bool $enable Flag indicating if the stacktrace must be auto-filled
+     * @param bool $enable flag indicating if the stacktrace will be attached to captureMessage calls
```
Comment (ste93cry) [inline src/Options.php]: Not that important after all, but the description of the parameter should always start with a capitalized letter

File: src/Options.php [hunk 10]
```diff
@@ -660,14 +614,10 @@ private function configureOptions(OptionsResolver $resolver)
...
             'sample_rate' => 1,
             'mb_detect_order' => null,
-            'auto_log_stacks' => true,
+            'attach_stacktrace' => false,
             'context_lines' => 3,
             'enable_compression' => true,
-            'current_environment' => 'default',
-            'environments' => [],
-            'excluded_loggers' => [],
-            'excluded_exceptions' => [],
```
Comment (ste93cry) [inline src/Options.php]: Not that important after all but I'm going to ask anyway: is there any reason regarding these options being moved to the end of the list?

File: src/Client.php (hunk 2 above)
Comment (HazAT) [reply src/Client.php]: We don't need this check, if there is an exception we always want the stacktrace.

Comment (HazAT) [reply src/Client.php]: `attachStacktrace` only determines if a stacktrace is attached to `captureMessage`

File: src/ClientBuilder.php (hunk 4 above)
Comment (HazAT) [reply src/ClientBuilder.php]: fixed

File: tests/ClientTest.php (hunk 5 above)
Comment (HazAT) [reply tests/ClientTest.php]: Explained above, for exceptions we always want a stacktrace.

File: tests/ClientBuilderTest.php (hunk 3 above)
Comment (HazAT) [reply tests/ClientBuilderTest.php]: Fixed

**Pull request merged**